Disable API key health checks on process/cluster agent

kevinconaway opened the pull request:
### What does this PR do?

The process/orchestrator intake does its own API key validation and does not implement the /validate endpoint that the forwarder health check is using. Having the health check enabled makes for a lot of noisy log lines in the agent

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

kevinconaway committed:
Disable API key health checks on process/cluster agent

The process/orchestrator intake does its own API key validation and does not implement the /validate endpoint that the forwarder health check is using. Having the health check enabled makes for a lot of noisy log lines in the agent

shang-wang commented on pkg/forwarder/forwarder.go:
```diff
@@ -166,16 +166,24 @@ type DefaultForwarder struct {
+	return NewDefaultForwarderWithOptions(keysPerDomains, true)
+}
+
+// NewDefaultForwarderWithOptions creates a new DefaultForwarder with the given configuration
```
I guess you are not going to use the esoteric way to set checker by returning a function that modifies the new service object? :)

hush-hush commented on pkg/forwarder/forwarder.go:
```diff
@@ -166,16 +166,24 @@ type DefaultForwarder struct {
+}
+
+// NewDefaultForwarderWithOptions creates a new DefaultForwarder with the given configuration
+func NewDefaultForwarderWithOptions(keysPerDomains map[string][]string, enableHealthChecks bool) *DefaultForwarder {
```
We should either have a `NewDefaultForwarderWithoutHealthCheck` that would call an internal function with `false` like `NewDefaultForwarder` does, or pass a options struct.

Right now we're not really passing options. Since we don't have that much option having too method `func NewDefaultForwarderWithOptions(keysPerDomains map[string][]string) *DefaultForwarder` and `func NewDefaultForwarder(keysPerDomains map[string][]string) *DefaultForwarder` should be fine.

WDYT ?

kevinconaway committed:
Use forwarder.Options for configuring DefaultForwarder

The commit modifies DefaultForwarder to use a forwarder.Options class to configure DefaultForwarder instead of relying on a mix of configuration values and constructor arguments

hush-hush commented on Gopkg.lock:
```diff
@@ -898,7 +898,7 @@
   version = "v1.0"
 
 [[projects]]
-  digest = "1:67cd1819192a5fed8a7378065f27cc4cec3022c36b3f0d6fface973a788cb642"
```
I think Gopkg.lock was commited by mistake, no ?

kevinconaway replied:
Yes, I'm not sure why this changed.

hush-hush reviewed (dismissed):
LGTM apart from the Gopkh.lock update.

Feel free to merge after fixing it

clamoriniere commented on pkg/forwarder/forwarder.go:
```diff
@@ -152,6 +152,24 @@ type Forwarder interface {
+	return &Options{
+		NumberOfWorkers:      config.Datadog.GetInt("forwarder_num_workers"),
+		RetryQueueSize:       config.Datadog.GetInt("forwarder_retry_queue_max_size"),
+		EnableHealthChecking: true,
```
I know that currently the Forwarder only check the API key validity. But I think that having only a On/Off for the HealthChecking is not futur proof.
what do you think to rename the option: `EnableHealthChecking` to `DisableAPIKeyChecking` which mean that by default the API Key check is enabled (backwardcompatibility)
```golang
type Options struct {
	NumberOfWorkers       int
	RetryQueueSize        int
	DisableAPIKeyChecking bool
	KeysPerDomain         map[string][]string
}
```
and forward the option to the `forwarderHealth{}` instantiation or a new method `NewForwarderHealth()`

```

func NewDefaultForwarder(options *Options) *DefaultForwarder {
	f := &DefaultForwarder{
		NumberOfWorkers:  config.Datadog.GetInt("forwarder_num_workers"),
		NumberOfWorkers:  options.NumberOfWorkers,
		domainForwarders: map[string]*domainForwarder{},
		keysPerDomains:   map[string][]string{},
		internalState:    Stopped,
		healthChecker:   NewForwarderHealth(options),
	}
```

kevinconaway replied:
Does it need to be future proof?  I don't believe this is intended to be used external to the agent.

clamoriniere replied:
when I say future proof. it is for example adding other check that only the API key validity, for example the size of a queue.
This PR deactivate completely the heath check, and introduce some logic to verify that the pointer is not nil.

clamoriniere commented on pkg/forwarder/forwarder.go:
```diff
@@ -152,6 +152,24 @@ type Forwarder interface {
+}
+
+// NewOptions creates new Options with default values
+func NewOptions(keysPerDomain map[string][]string) *Options {
```
Could you rename this function to `NewDefaultOptions` because this function didn't allow the user to set all the possible options.

gzussa reviewed (approved):
LGTM
